src/gpl/src/initialPlace.cpp

```diff
@@ -106,23 +142,157 @@ void InitialPlace::doBicgstabPlace() {
 
   // set ExtId for idx reference // easy recovery
   setPlaceInstExtId();
+
+  // Parameters that don't change with iteration
+  int m = pb_->placeInsts().size(); // number of rows of matrix A
+  float tol = 1e-6; // Threshold to decide if the 
+  int reorder = 0;  // "0" for common matrix meaning no extra processing
+  int singularity = -1; // Output. Showing if the sparse matrix is signular or not. -1 = A is invertible
+
   for(int i=1; i<=ipVars_.maxIter; i++) {
     updatePinInfo();
     createSparseMatrix();
+    if (GPU == 1){
+      // Set sparse matrices from dense matrices.
+      cooRowIndexX.clear();
+      cooColIndexX.clear();
+      cooValX.clear();
+      cooRowIndexY.clear();
+      cooColIndexY.clear();
+      cooValY.clear();
+      for (size_t i = 0; i < placeInstForceMatrixX_.rows(); i ++){
+        for (size_t j = 0; j < placeInstForceMatrixX_.cols(); j ++){
+          if (placeInstForceMatrixX_.coeffRef(i,j) != 0){
+            cooRowIndexX.push_back(i);
+            cooColIndexX.push_back(j);
+            cooValX.push_back(placeInstForceMatrixX_.coeffRef(i,j));
+          } 
+          if (placeInstForceMatrixY_.coeffRef(i,j) != 0){
+            cooRowIndexY.push_back(i);
+            cooColIndexY.push_back(j);
+            cooValY.push_back(placeInstForceMatrixY_.coeffRef(i,j));
+          } 
+        }
+      }
 
-    // BiCGSTAB solver for initial place
-    BiCGSTAB< SMatrix, IdentityPreconditioner > solver;
-    solver.setMaxIterations(ipVars_.maxSolverIter);
-    solver.compute(placeInstForceMatrixX_);
-    instLocVecX_ = solver.solveWithGuess(fixedInstForceVecX_, instLocVecX_);
-    errorX = solver.error();
+      int nnz_x = cooRowIndexX.size(); // number of non-zeros for matrix X
+      int nnz_y = cooRowIndexY.size(); // number of non-zeros for matrix Y
+
+      // Allocate device memeory and copy data to device
+      int *d_cooRowIndexX, *d_cooColIndexX,  *d_cooRowIndexY, *d_cooColIndexY;
+      float *d_cooValX, *d_instLocVecX, *d_fixedInstForceVecX_, *d_cooValY, *d_instLocVecY, *d_fixedInstForceVecY_;
+      CUDA_ERROR(cudaMalloc((void**)&d_cooRowIndexX, nnz_x * sizeof(int)));
+      CUDA_ERROR(cudaMalloc((void**)&d_cooColIndexX, nnz_x * sizeof(int)));
+      CUDA_ERROR(cudaMalloc((void**)&d_cooValX, nnz_x * sizeof(float)));
+      CUDA_ERROR(cudaMalloc((void**)&d_fixedInstForceVecX_, m * sizeof(float)));
+      CUDA_ERROR(cudaMalloc((void**)&d_instLocVecX, m * sizeof(float)));
+
+      CUDA_ERROR(cudaMalloc((void**)&d_cooRowIndexY, nnz_y * sizeof(int)));
+      CUDA_ERROR(cudaMalloc((void**)&d_cooColIndexY, nnz_y * sizeof(int)));
+      CUDA_ERROR(cudaMalloc((void**)&d_cooValY, nnz_y * sizeof(float)));
+      CUDA_ERROR(cudaMalloc((void**)&d_fixedInstForceVecY_, m * sizeof(float)));
```

Comment:
Any failures in this will lead to memory leaks if not inconsistent state.  This is very old C style programming.  If the cuda library forces this then you should be safe abstractions on top of it or find a better library.

Reply:
I see. From my knowledge, this is the only way. If I want to secure the command, what I can think of is to "Free all CUDA variables" once the program gets an unexpected problem and exits. I will also discuss with Chris about possible solution for it.

Reply:
@RuijiaoSun the alternative to the macro is turning them into c++ functions which would be more well received.